### PR TITLE
Audit 0.2.1

### PR DESCRIPTION
**Pull request by marco-sundsk**
# Ref exchange audit

1. MAX_ACCOUNT_LENGTH is incorrect -- fixed;
2. u128 can not be correctly serialized -- fixed;
3. `execute_actions` should verify actions number -- reject to enable nope action;
4. `execute_actions` is not payable -- fixed;
5. Wrong error code in `execute_actions` -- fixed;
6. `parse_token_id` logic may cause problem. -- fixed;
> using `:nn` as a partern of pool lpt token
7. `mft_transfer` may cause problem. -- fixed;
8. Use different spec for MFT -- fixed;
9. `migrate` is not private -- fixed;
10. Simple pool can be created with `0` tokens. -- fixed;
11. LookupMap `shares` prefix issue -- fixed with StorageKey;
12. Ability to drain the pool without paying using instant swap. -- fixed;
13. Never unregister a LP when he remove liqudity. 
14. shutdown instant swap from the interface;

### CODE REVIEW

File: ref-exchange/src/lib.rs
Comment (evgenykuzyakov) [inline ref-exchange/src/lib.rs:173]: This methods should be public, but maybe marked as `#[payable]`.

File: ref-exchange/src/owner.rs
Comment (evgenykuzyakov) [inline ref-exchange/src/owner.rs:39]: @referencedev Just to confirm that we call `migrate` in the same receipt as we deploy new contract.

**Review by evgenykuzyakov — APPROVED** [review]
Looks good to me. Please wait for @referencedev review before merging.

File: ref-exchange/src/errors.rs
Comment (referencedev) [inline ref-exchange/src/errors.rs:19]: I think if we want to be super clear this should be "Attach 1 yN to swap tokens not in whitelist" or something. Deposit is a bit overloaded in this context. There is value in saying that access key is not allowed - because the attached deposit is indeed to prevent access keys to doing this operation.

File: ref-exchange/src/multi_fungible_token.rs
Comment (referencedev) [inline ref-exchange/src/multi_fungible_token.rs:38]: I'm not sure where NFT standard end up, but one option is to use the ":" symbol - it was similar idea that it's no allowed in the accounts and it allows to create "deep" link into something inside a contract: "contract_name:token_id". Not sure it's useful to define this here and using something is better than nothing.

File: ref-exchange/src/owner.rs
Comment (referencedev) [reply ref-exchange/src/owner.rs:39]: It's done in the same call, yes. See code below of `upgrade` where it's initiated.

File: ref-exchange/src/simple_pool.rs
Comment (referencedev) [inline ref-exchange/src/simple_pool.rs:54]: let's rename this to NUM_TOKENS? as there is only 2 allowed either way.

File: ref-farming/Cargo.toml
Comment (referencedev) [inline ref-farming/Cargo.toml:11]: let's remove this?

File: ref-exchange/src/account_deposit.rs
Comment (referencedev) [inline ref-exchange/src/account_deposit.rs:47]: Hm, let's add 4 to MAX_ACCOUNT_LENGTH and rename to `MAX_ACCOUNT_BYTES`. 
Because it's used in other places. Like in `MIN_ACCOUNT_DEPOSIT_LENGTH` which also should be `MAX_ACCOUNT_BALANCES_BYTES`.

File: ref-exchange/src/simple_pool.rs
Comment (referencedev) [inline ref-exchange/src/simple_pool.rs:187]: let's add some unit tests that would
 a) fail with previous solution
 b) fail if this string is commented

File: ref-exchange/src/multi_fungible_token.rs
Comment (referencedev) [inline ref-exchange/src/multi_fungible_token.rs:70]: let's add a test for this.

**Review by referencedev — APPROVED** [review]
Overall looks good with few comments.

Few comments:
 - If we already redoing the state layout, let's add `storage_used` variable to `Account`. This is something needed for proper handling of LP tokens storage, but wasn't added before of complications with migration. (see #7 and #27 as a patch)
 - Even better add Versioning around `Account` now so we can actually upgrade it in the future. (also see #7)

File: ref-exchange/Cargo.toml
Comment (referencedev) [inline ref-exchange/Cargo.toml:13]: also there is 4.0 of sdk now, isn't it?
There is PR for this in #34

Comment (evgenykuzyakov) [reply ref-exchange/Cargo.toml:13]: I think `4.0.0` is not published yet. Only the preview version. Better wait some time before upgrading.

Comment (marco-sundsk) [reply ref-exchange/Cargo.toml:13]: OK, I also prefer to keeping 3.1.0 for safety.

File: ref-exchange/src/account_deposit.rs
Comment (marco-sundsk) [reply ref-exchange/src/account_deposit.rs:47]: It's a little tricky that  in `MIN_ACCOUNT_DEPOSIT_LENGTH` the `MAX_ACCOUNT_LENGTH`  is supposed to be the key length of a LookupMap item, which has an extra prefix compare to HashMap item in the Account Structure. It's more like a `4 bytes + len(prefix)`, but I am really uncertain how to code that.

File: ref-farming/Cargo.toml
Comment (marco-sundsk) [reply ref-farming/Cargo.toml:11]: OK

File: ref-exchange/src/simple_pool.rs
Comment (marco-sundsk) [reply ref-exchange/src/simple_pool.rs:187]: OK

File: ref-exchange/src/multi_fungible_token.rs
Comment (marco-sundsk) [reply ref-exchange/src/multi_fungible_token.rs:70]: OK

File: ref-exchange/src/simple_pool.rs
Comment (marco-sundsk) [reply ref-exchange/src/simple_pool.rs:54]: OK

File: ref-exchange/src/errors.rs
Comment (marco-sundsk) [reply ref-exchange/src/errors.rs:19]: OK

File: ref-exchange/src/multi_fungible_token.rs
Comment (marco-sundsk) [reply ref-exchange/src/multi_fungible_token.rs:38]: changed from `#` to `:`

File: ref-exchange/src/owner.rs
Comment (marco-sundsk) [reply ref-exchange/src/owner.rs:39]: So, we are good here, right?

File: ref-exchange/src/account_deposit.rs
Comment (evgenykuzyakov) [inline ref-exchange/src/account_deposit.rs:15]:
```suggestion
// [AUDIT_01]
const MAX_ACCOUNT_LENGTH: u128 = 64;
const MAX_ACCOUNT_BYTES: u128 = MAX_ACCOUNT_LENGTH + 4;
const MIN_ACCOUNT_DEPOSIT_LENGTH: u128 = MAX_ACCOUNT_LENGTH + 16 + 4;
```

File: ref-exchange/src/account_deposit.rs
Comment (evgenykuzyakov) [inline ref-exchange/src/account_deposit.rs]:
```suggestion
        (MIN_ACCOUNT_DEPOSIT_LENGTH + self.tokens.len() as u128 * (MAX_ACCOUNT_BYTES + 16))
```

**Review by evgenykuzyakov — APPROVED** [review]

**Pull request merged**
